include/exec/repeat_effect_until.hpp

```diff
@@ -0,0 +1,211 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "../stdexec/execution.hpp"
+#include "__detail/__manual_lifetime.hpp"
+#include "stdexec/concepts.hpp"
+#include <concepts>
+
+namespace exec {
+  namespace __repeat_effect_until {
+    using namespace stdexec;
+
+    template <typename _SourceId, typename _ReceiverId>
+    struct __receiver {
+      struct __t;
+    };
+
+    template <typename _SourceId, typename _ReceiverId>
+    struct __operation {
+      using _Source = stdexec::__t<_SourceId>;
+      using _Receiver = stdexec::__t<_ReceiverId>;
+
+      struct __t : stdexec::__immovable {
+        using __id = __operation;
+        using __receiver_t = stdexec::__t<__receiver<_SourceId, _ReceiverId>>;
+        using __source_op_t = stdexec::connect_result_t<_Source, __receiver_t>;
+
+        [[no_unique_address]] _Source __source_;
+        [[no_unique_address]] _Receiver __receiver_;
+        __manual_lifetime<__source_op_t> __source_op_;
+
+        template <typename _Source2, typename _Receiver2>
+        __t(_Source2 &&__source, _Receiver2 &&__receiver) noexcept(
+          std::is_nothrow_constructible_v<_Source, _Source2>
+            &&std::is_nothrow_constructible_v<_Receiver, _Receiver2>
+              &&__nothrow_connectable<_Source, __receiver_t>)
+          : __source_((_Source2 &&) __source)
+          , __receiver_((_Receiver2 &&) __receiver) {
+          __source_op_.__construct_with([&] {
+            return stdexec::connect((_Source2 &&) __source_, __receiver_t{this});
+          });
+        }
+
+        ~__t() {
+          __source_op_.__destruct();
+        }
+
+        friend void tag_invoke(stdexec::start_t, __t &__self) noexcept {
+          stdexec::start(__self.__source_op_.__get());
+        }
+      };
+    };
+
+    template <typename _SourceId, typename _ReceiverId>
+    struct __receiver<_SourceId, _ReceiverId>::__t {
+      using __id = __receiver;
+      using _Source = stdexec::__t<_SourceId>;
+      using _Receiver = stdexec::__t<_ReceiverId>;
+      using __op_t = stdexec::__t<__operation<_SourceId, _ReceiverId>>;
+
+      __op_t *__op_;
+
+      explicit __t(__op_t *op) noexcept
+        : __op_(op) {
+      }
+
+      __t(__t &&__other) noexcept
+        : __op_(std::exchange(__other.__op_, {})) {
+      }
+
+      template <__decays_to<__t> _Self, convertible_to<bool> _Arg>
+      friend void tag_invoke(set_value_t, _Self &&__self, _Arg &&__flag) noexcept {
+        __self.__op_->__source_op_.__destruct();
```

Comment:
Here you destroy the operation that owns this `receiver::__t` object, right? This could be the culprit in ASAN. If I am right, using subsequent `__self.__op` would trigger it.

Reply:
Yes, thank you for pointing out the problem. Calling the destructor here does cause __self to lose validity